tests/nested/manual/core20-set-efi-boot-vars/task.yaml

```diff
@@ -0,0 +1,206 @@
+    # manually when executing efibootmgr.
+    remote_exec_efibootmgr() {
+        remote_path="$(remote.exec "echo $PATH")"
+        augmented_path="/snap/toolbox/current/bin:/snap/toolbox/current/sbin:/snap/toolbox/current/usr/bin:/snap/toolbox/current/usr/sbin:${remote_path}"
+        remote.exec "PATH=${augmented_path} LD_LIBRARY_PATH=/lib:/usr/lib:/lib/x86_64-linux-gnu:/usr/lib/x86_64-linux-gnu:/snap/toolbox/current/lib:/snap/toolbox/current/usr/lib:/snap/toolbox/current/lib/x86_64-linux-gnu:/snap/toolbox/current/usr/lib/x86_64-linux-gnu sh -c 'efibootmgr -v'"
+    }
+
+    # Check that the EFI assets are correctly situated
+    check_efi_assets() {
+        if [ "$1" = "boot" ]; then
+            remote.exec "ls /run/mnt/ubuntu-seed/EFI/boot/boot${ARCH}.efi"
+            remote.exec "ls /run/mnt/ubuntu-seed/EFI/boot/grub${ARCH}.efi"
+        elif [ "$1" = "ubuntu" ]; then
+            remote.exec "ls /run/mnt/ubuntu-seed/EFI/ubuntu/BOOT${ARCH_UPPER}.CSV"
+            remote.exec "ls /run/mnt/ubuntu-seed/EFI/ubuntu/grub${ARCH}.efi"
+            remote.exec "ls /run/mnt/ubuntu-seed/EFI/ubuntu/shim${ARCH}.efi"
+            remote.exec "ls /run/mnt/ubuntu-seed/EFI/boot/boot${ARCH}.efi"
+            remote.exec "ls /run/mnt/ubuntu-seed/EFI/boot/fb${ARCH}.efi"
+        fi
+    }
+
+    # Print the number of the first boot option in BootOrder.
+    get_bootorder_first_number() {
+        remote_exec_efibootmgr | MATCH "^BootOrder: "
+        remote_exec_efibootmgr | grep  "^BootOrder: " | cut -d ' ' -f 2 | cut -d ',' -f 1
+    }
+    # Check that a boot option variable exists for the given shim path, and
+    # that it is the first option in the BootOrder variable.
+    check_efi_variables() {
+        shim_path="$(echo "$1" | sed 's/\//\\\\/g')"
+        boot_var_pattern="^Boot[0-9A-F][0-9A-F][0-9A-F][0-9A-F]. ubuntu.*/File.${shim_path}.$"
+        echo "Check that there exists one EFI boot variable for ${shim_path}"
+        remote_exec_efibootmgr | MATCH "${boot_var_pattern}"
+        remote_exec_efibootmgr | grep -c "${boot_var_pattern}" | MATCH "^1$"
+        echo "Get boot number associated with that boot variable"
+        boot_num=$(remote_exec_efibootmgr | grep "${boot_var_pattern}" | grep -o '^Boot....' | sed 's/Boot//')
+        echo "Check that first boot option in BootOrder matches ${boot_num}"
+        get_bootorder_first_number | MATCH "^${boot_num}$"
+    }
+
+    echo "Check that EFI assets in EFI/${INITIAL_ASSET_DIR} and variables are set correctly"
+    check_efi_assets "$INITIAL_ASSET_DIR"
+    if [ "${INITIAL_ASSET_DIR}" = "boot" ]; then
+        check_efi_variables "/EFI/boot/boot${ARCH}.efi"
+    elif [ "${INITIAL_ASSET_DIR}" = "ubuntu" ]; then
+        check_efi_variables "/EFI/ubuntu/shim${ARCH}.efi"
+    fi
+
+    echo "All good!"
+
+    # At the moment, setting EFI boot vars does not happen during gadget update.
+    # When this is fixed, include the rest of this test.
```

Comment:
Since we'll build the gadget to have assets in a different location, we know the new variable will be different than the old one, so there shouldn't be any conflict. The EFI variables code doesn't remove old boot variables, even if they no longer point to anything meaningful. I don't think we need to clean up the old variables? The only conflict I see is if we used the same description for both, but we grep for a pattern matching a given shim path as well, so I don't think it should be a problem.

Reply:
I was thinking here that you wanted to test updating by re-installing the same gadget that you built the image with. In that case you would need to clean-up the boot variables to check that on gadget updates they are set again.

Note that I am talking about cleaning up the variables in the test by using `efibootmgr`, not about using snapd code for that, which we do not need.

Reply:
This would be testing that updating the gadget to a system that was installed with an old snapd version (which would not have set EFI variables) updates the EFI variables.

Reply:
Ah, that would be a good test to have. I assume that gadget edition would still need to be higher than the previous? I'll add a variant that installs a gadget with `EFI/ubuntu/shimx64.efi`, deletes the Boot#### variable it creates, and then reinstalls an identical gadget with a bumped edition number. Is there anything else special you think a test like this (emulating somewhat an older snapd install, then updating gadget) would need?

Reply:
Actually the content itself needs a different hash besides the edition number, so maybe some empty file could be added to the contents, for instance. Anyway this test would be for a follow-up.